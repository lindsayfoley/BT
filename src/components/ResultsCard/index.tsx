import { FunctionComponent } from 'react';
interface ISource {
  id: string;
  name: string;
}
export interface IArticle {
  title: string;
  author: string;
  description: string;
  publishedAt: string;
  url: string;
  source: ISource;
}

export const CHARACTER_LIMIT = 150;
export const truncateText = (text: string) => `${text.slice(0, CHARACTER_LIMIT)}...`;

const ResultsCard: FunctionComponent<IArticle> = (article: IArticle) => {
  const {
    title,
    author,
    description,
    publishedAt,
    url,
    source: { name, id },
  } = article;

  const truncatedDescription = description && truncateText(description);

  return (
    <>
      <article key={publishedAt}>
        <h3>{title}</h3>
        <span className="author">{author || name || id}</span>
        <p>
          <span className="description" dangerouslySetInnerHTML={{ __html: truncatedDescription }} />
          <a target="_blank" rel="noreferrer" href={url}>
            read more &gt;
          </a>
        </p>
      </article>
      <hr />
    </>
  );
};

export default ResultsCard;
