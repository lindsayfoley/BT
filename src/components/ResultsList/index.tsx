import { FunctionComponent } from 'react';
import ResultsCard, { IArticle } from '../ResultsCard';

export interface IResultsListProps {
  list: IArticle[];
}

const ResultsList: FunctionComponent<IResultsListProps> = ({ list }) => {
  if (!list || list.length === 0) {
    return null;
  }

  return (
    <section>
      <h2>{`Showing ${list.length} results:`}</h2>
      {list.map((article, index) => (
        <ResultsCard {...article} key={`${article.title}-${index}`} />
      ))}
    </section>
  );
};

export default ResultsList;
