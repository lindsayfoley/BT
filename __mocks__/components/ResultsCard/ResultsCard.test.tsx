import React from 'react';
import { shallow, render} from 'enzyme';
import ResultsCard, { IArticle, CHARACTER_LIMIT, truncateText } from '../../../src/components/ResultsCard';
import { mockArticleList } from '../ResultsList/ResultsList.test';

const mockArticle = mockArticleList[0];

const setup = (useRender: boolean = false) => {
  if (useRender) {
    return render(<ResultsCard {...mockArticle} />);
  }

  return shallow(<ResultsCard {...mockArticle} />);
};

describe('<ResultsCard />', () => {
  let wrapper;
  
  it('renders without crashing', () => {
    wrapper = setup();
    expect(wrapper.isEmptyRender()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it(`should truncate the description to ${CHARACTER_LIMIT} words and add an ellipsis`, () => {
    wrapper = setup(true);
    expect(wrapper.find('.description').text()).toContain(truncateText(mockArticle.description));
    expect(wrapper.find('.description').text()).toContain('...');
  });

  it('should render an article with all required details', () => {
    wrapper = setup();
    expect(wrapper.find('article')).toHaveLength(1);
    expect(wrapper.find('h3').text()).toContain(mockArticle.title);
    expect(wrapper.find('.author').text()).toContain(mockArticle.author);
    expect(wrapper.find('.description')).toHaveLength(1);
    expect(wrapper.find('a').last().prop('href')).toContain(mockArticle.url);
    expect(wrapper.find('hr')).toHaveLength(1);
  });
});
