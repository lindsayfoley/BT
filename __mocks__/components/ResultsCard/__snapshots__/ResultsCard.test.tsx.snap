// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsCard /> renders without crashing 1`] = `
<Fragment>
  <article
    key="2021-02-17T09:30:00-07:00"
  >
    <h3>
      An earnest review of a robotic cat pillow
    </h3>
    <span
      className="author"
    >
      Brian Heater
    </span>
    <p>
      <span
        className="description"
        dangerouslySetInnerHTML={
          Object {
            "__html": "You don’t need Qoobo in your life. Nobody needs Qoobo, exactly. In fact, first reactions tend to range from befuddlement to bemusement. The robotic ca...",
          }
        }
      />
      <a
        href="https://techcrunch.com/2020/12/30/an-earnest-review-of-a-robotic-cat-pillow/"
        rel="noreferrer"
        target="_blank"
      >
        read more &gt;
      </a>
    </p>
  </article>
  <hr />
</Fragment>
`;

exports[`<ResultsList /> renders without crashing 1`] = `
<section
  className="results"
>
  <h2>
    Showing 2 results:
  </h2>
  <ResultsCard
    author="Brian Heater"
    description="You don’t need Qoobo in your life. Nobody needs Qoobo, exactly. In fact, first reactions tend to range from befuddlement to bemusement. The robotic cat pillow doesn’t make a ton of sense on the face of it — in part because Qoobo has no face."
    key="2021-02-17T09:30:00-07:00"
    publishedAt="2021-02-17T09:30:00-07:00"
    source={
      Object {
        "id": "123456",
        "name": "Techcrunch",
      }
    }
    title="An earnest review of a robotic cat pillow"
    url="https://techcrunch.com/2020/12/30/an-earnest-review-of-a-robotic-cat-pillow/"
  />
  <ResultsCard
    author="BBC News"
    description="Tony Wittman allegedly stormed the shelter on Monday in full camouflage gear and pointed an assault rifle at a female worker."
    key="2021-01-17T09:30:00-05:00"
    publishedAt="2021-01-17T09:30:00-05:00"
    source={
      Object {
        "id": "789101",
        "name": "BBC News",
      }
    }
    title="Australia: Man held after armed raid in search of cat"
    url="https://www.bbc.co.uk/news/world-australia-55667023"
  />
</section>
`;
